vowpalwabbit/slim/include/vw_slim_predict.h

```diff
@@ -13,6 +13,150 @@
 #include "model_parser.h"
 #include "opts.h"
 
+namespace vw_slim { namespace internal {
+  template <typename It1, typename It2>
+  class location_reference;
+
+  template <typename It1, typename It2>
+  class location_value
+  {
+   public:
+    using Val1 = typename std::iterator_traits<It1>::value_type;
+    using Val2 = typename std::iterator_traits<It2>::value_type;
+
+    Val1 _val1;
+    Val2 _val2;
+
+    location_value(
+        const location_reference<typename It1, typename It2>& rhs)
+    : _val1(*rhs._ptr1), _val2(*rhs._ptr2) {}
+  };
+
+  template <typename It1, typename It2>
+  class location_reference
+  {
+   public:
+    It1 _ptr1;
+    It2 _ptr2;
+    using Ref = location_reference<It1, It2>;
+    using Val = location_value<It1, It2>;
+
+    location_reference() = delete;
+
+    location_reference(const Ref& other)
+    : _ptr1(other._ptr1), _ptr2(other._ptr2) {}
+
+    location_reference(It1 first, It2 second)
+    : _ptr1(first), _ptr2(second) {}
+
+    location_reference& operator=(const Val& rhs)
+    {
+      *_ptr1 = rhs._val1;
+      *_ptr2 = rhs._val2;
+      return *this;
+    }
+
+    location_reference& operator=(const Ref& rhs)
+    {
+      *_ptr1 = *rhs._ptr1;
+      *_ptr2 = *rhs._ptr2;
+      return *this;
+    }
+  };
+
+  template <typename It1, typename It2>
+  class collection_pair_iterator
+      : public std::iterator<std::random_access_iterator_tag,
+            location_value<It1, It2>,         // value type
+            size_t,                           // difference type
+            location_reference<It1, It2>      // reference_type
+            >
+  {
+    It1 _ptr1;
+    It2 _ptr2;
+
+   public:
+    using Iter = collection_pair_iterator<It1, It2>;
+    using Loc = location_value<It1, It2>;
+    using Ref = location_reference<It1, It2>;
+
+    collection_pair_iterator(It1 first, It2 second) : _ptr1(first), _ptr2(second) {}
+
+    // must support: a) default-construction, b) copy-construction, c) copy-assignment, d) destruction
+    collection_pair_iterator() = default;
+    collection_pair_iterator(const collection_pair_iterator& rhs) = default;
+    collection_pair_iterator& operator=(const collection_pair_iterator& rhs) = default;
+
+    // must support: a == b; a != b;
+    bool operator==(const Iter& rhs) const { return (_ptr1 == rhs._ptr1 && _ptr2 == rhs._ptr2); }
+    bool operator!=(const Iter& rhs) const { return !(operator==(rhs)); }
+
+    // must support: b = *a; a->m ;
+    // must support: *a = t;
+    Ref operator*() { return Ref(_ptr1, _ptr2); }  // Non-conforming - normally returns loc&
+
+    // must support: ++a; a++; *a++;
+    Iter& operator++()
+    {
+      ++_ptr1;
```

Comment:
What stops this going past the bounds?

Reply:
Represents an iterator that is not aware of bounds.  Bounds are managed by std::sort()

---
In reply to: [310806653](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1800#discussion_r310806653) [](ancestors = 310806653)